### PR TITLE
feat: [PIE-3422]: added api pattern to api telemetry

### PR DESCRIPTION
**Pull request by vaibhav-si**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/32137)

### CODE REVIEW

**Review by brijeshdhakar — DISMISSED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32137)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Review by rijajoo — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32137)* status: 4 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[950-telemetry/src/main/java/io/harness/telemetry/utils/TelemetryDataUtils.java, line 87 at r1](https://reviewable.io/reviews/harness/harness-core/32137#-N-3JBz3-0mFSUOVd1H8:-N-3JBz3-0mFSUOVd1H9:bqoh1m1) ([raw file](https://github.com/harness/harness-core/blob/a5b82ab5ff133352881fda5fe85dbed67884a247/950-telemetry/src/main/java/io/harness/telemetry/utils/TelemetryDataUtils.java#L87)):*
> ```Java
>     StringBuilder pattern = new StringBuilder("");
>     for (int i = 0; i < templates.size(); i++) {
>       pattern.append(templates.get(templates.size() - 1 - i).getTemplate());
> ```

Why do we reverse the matched templates?

**Comment by brijeshdhakar** [comment]
*[950-telemetry/src/main/java/io/harness/telemetry/utils/TelemetryDataUtils.java, line 87 at r1](https://reviewable.io/reviews/harness/harness-core/32137#-N-3JBz3-0mFSUOVd1H8:-N-3vJRu7ro_PbYvLWjN:b-7kdbd6) ([raw file](https://github.com/harness/harness-core/blob/a5b82ab5ff133352881fda5fe85dbed67884a247/950-telemetry/src/main/java/io/harness/telemetry/utils/TelemetryDataUtils.java#L87)):*
<details><summary><i>Previously, rijajoo (Richa Jajoo) wrote…</i></summary><blockquote>

Why do we reverse the matched templates?
</blockquote></details>

This list is in order of matched method level template and then class level template. So to create the complete matched API we need to reverse the order.
I tested this in API Errors telemetry event and it works correctly this way.

**Review by rijajoo — DISMISSED** [review]
:lgtm:

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32137#-:-N-MJRVb5X1zLmXxriA4:bnfp4nl)* status: Some of the checks are too old.

**Review by rijajoo — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32137#-:-N-MjUnVBYTQMmkllGer:bnfp4nl)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32137)* status: :shipit: complete! ✓

File: 950-telemetry/src/main/java/io/harness/telemetry/utils/TelemetryDataUtils.java [hunk 1]
```diff
@@ -75,4 +79,13 @@ public static Map<String, String> filterNonNullProperties(Map<String, Object> pr
     }
     return nonNullProperties;
   }
+
+  public static String getApiPattern(ContainerRequestContext containerRequestContext) {
+    List<UriTemplate> templates = ((UriRoutingContext) containerRequestContext.getUriInfo()).getMatchedTemplates();
+    StringBuilder pattern = new StringBuilder("");
+    for (int i = 0; i < templates.size(); i++) {
```
[the remark this answers is not in the document]
Comment (vaibhav-si) [reply 950-telemetry/src/main/java/io/harness/telemetry/utils/TelemetryDataUtils.java:86]: I checked the library code and it never returns null.

File: 950-telemetry/src/main/java/io/harness/telemetry/utils/TelemetryDataUtils.java [hunk 2]
```diff
@@ -75,4 +79,13 @@ public static Map<String, String> filterNonNullProperties(Map<String, Object> pr
...
+      pattern.append(templates.get(templates.size() - 1 - i).getTemplate());
```
[the remark this answers is not in the document]
Comment (vaibhav-si) [reply 950-telemetry/src/main/java/io/harness/telemetry/utils/TelemetryDataUtils.java:87]: yes, its documented. I showed to Richa the same yesterday

**Pull request merged**